core/src/datayoga_core/blocks/relational/write/block.py

```diff
@@ -18,49 +19,75 @@
 
 
 class Block(DyBlock, metaclass=ABCMeta):
+    _engine_fields = ("business_key_columns", "mapping_columns", "columns",
+                      "delete_stmt", "upsert_stmt", "tbl", "connection", "engine")
 
     def init(self, context: Optional[Context] = None):
         logger.debug(f"Initializing {self.get_block_name()}")
 
-        self.engine, self.db_type = relational_utils.get_engine(self.properties.get("connection"), context)
+        self.context = context
+        self.engine = None
+        self.setup_engine()
 
-        self.schema = self.properties.get("schema")
-        self.table = self.properties.get("table")
-        self.opcode_field = self.properties.get("opcode_field")
-        self.load_strategy = self.properties.get("load_strategy")
-        self.keys = self.properties.get("keys")
-        self.mapping = self.properties.get("mapping")
-        self.foreach = self.properties.get("foreach")
+    def setup_engine(self):
+        if self.engine:
+            return
 
-        self.tbl = sa.Table(self.table, sa.MetaData(schema=self.schema), autoload_with=self.engine)
+        try:
+            engine, self.db_type = relational_utils.get_engine(self.properties.get("connection"), self.context)
 
-        logger.debug(f"Connecting to {self.db_type}")
-        self.connection = self.engine.connect()
+            logger.debug(f"Connecting to {self.db_type}")
+            connection = engine.connect()
 
-        if self.db_type in (relational_utils.DbType.SQLSERVER, relational_utils.DbType.ORACLE):
-            # MERGE statement requires this
-            self.connection = self.connection.execution_options(autocommit=True)
+            self.schema = self.properties.get("schema")
+            self.table = self.properties.get("table")
+            self.opcode_field = self.properties.get("opcode_field")
+            self.load_strategy = self.properties.get("load_strategy")
+            self.keys = self.properties.get("keys")
+            self.mapping = self.properties.get("mapping")
+            self.foreach = self.properties.get("foreach")
+            tbl = sa.Table(self.table, sa.MetaData(schema=self.schema), autoload_with=engine)
 
-        if self.opcode_field:
-            self.business_key_columns = [column["column"] for column in write_utils.get_column_mapping(self.keys)]
-            self.mapping_columns = [column["column"] for column in write_utils.get_column_mapping(self.mapping)]
+            self.engine = engine
+            self.connection = connection
```

Comment:
I think those should be declared after line 40 (the connection created successfully) for better readability.

Reply:
✅